### PR TITLE
MINOR:Upgrade guide updates for KIP-479

### PR DESCRIPTION
**Pull request by bbejeck**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bbejeck** [comment]
ping @mjsax, @vvcephei, and @ableegoldman 

\cc @JimGalasyn @londoncalling

File: docs/streams/upgrade-guide.html [hunk 1]
```diff
@@ -76,8 +76,13 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
 
     <!-- Placeholder KIP-213 -->
     <!-- Placeholder KIP-307 -->
-    <!-- Placeholder KIP-479 -->
+    <p>
+        Before the 2.4.0 version of Kafa Streams, users of the DSL could not name the state stores involved in a Stream-Stream join.  If users added DSL operations upstream of the
```
Comment (ableegoldman) [inline docs/streams/upgrade-guide.html]: nit: `stream-stream` join (or else `KStream-KStream`) here and below

File: docs/streams/upgrade-guide.html [hunk 2]
```diff
@@ -76,8 +76,13 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
...
+        object, which gives users the ability to name the join processor, repartition topic(s) (if a repartition is required), and the state stores involved in the join.  Also, by naming the state stores, the changelog topics backing the state stores are named as well.
 
+        Another feature delivered by <code>StreamJoined</code> that will be of interest to users is that you can now configure the type of state store used in the join.  Users can now elect to use in-memory stores or custom state stores for a Stream-Stream join.  Users should note that the provided stores will not be available for searching via Interactive Queries.  With the addition of <code>StreamJoined</code> Stream-Stream join operations using <code>Joined</code> have been deprecated, and users are encouraged to switch over to Stream-Stream join methods using the new overloaded methods.  You can get more details from <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-479%3A+Add+StreamJoined+config+object+to+Join">KIP-479</a>
```
Comment (ableegoldman) [inline docs/streams/upgrade-guide.html]: nits: (single line so can't refer to specific location)
1) "searching via IQ" -> "querying via IQ"? 
2) The sentence beginning "With the addition..." reads a bit awkwardly to me. Maybe something like "stream-stream joins accepting a Joined parameter have been deprecated; users are encouraged to migrate to the new StreamJoined parameter (is that an accurate description?)
3) I would also maybe refactor the first sentence a bit, something like "With the addition of the StreamJoined object, you can now configure the type of state store..."

**Review by ableegoldman — APPROVED** [review]
LGTM, left a few suggestions but feel free to take or leave any/all of them

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]:
```suggestion
        Before the 2.4.0 version of Kafa Streams, users of the DSL could not name the state stores involved in a Stream-Stream join.  If users added DSL operations before the 
```

Super subtle difference, but the only thing that affects the names is the order in which you call the DSL operators, so you can actually add upstream operations, as long as you add the new DSL calls after the original ones. I wouldn't get into it in the docs, but it's easy to swap "before" for "upstream" and be technically correct.

File: docs/streams/upgrade-guide.html [hunk 3]
```diff
@@ -76,8 +76,13 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
 
     <!-- Placeholder KIP-213 -->
     <!-- Placeholder KIP-307 -->
-    <!-- Placeholder KIP-479 -->
+    <p>
+        Before the 2.4.0 version of Kafa Streams, users of the DSL could not name the state stores involved in a Stream-Stream join.  If users added DSL operations upstream of the 
+        join, the internal names of the state stores would shift, requiring an application reset when redeploying.  Now in the 2.4.0 release Kafka Streams adds the <code>StreamJoined</code>
```
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]:
```suggestion
        join, the internal names of the state stores would shift, requiring an application reset when redeploying.  In the 2.4.0 release, Kafka Streams adds the <code>StreamJoined</code>
```

File: docs/streams/upgrade-guide.html (hunk 2 above)
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]:
```suggestion
        Another feature delivered by <code>StreamJoined</code> is that you can now configure the type of state store used in the join.  You can now elect to use in-memory stores or custom state stores for a Stream-Stream join.  Note that the provided stores will not be available for searching via Interactive Queries.  With the addition of <code>StreamJoined</code>, Stream-Stream join operations using <code>Joined</code> have been deprecated. Please switch over to Stream-Stream join methods using the new overloaded methods.  You can get more details from <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-479%3A+Add+StreamJoined+config+object+to+Join">KIP-479</a>
```

Several suggestions for this line:
* The audience of this documentation is the user, so it makes more sense to say "you" than "the user". If you look back over it, you'll see we mix both forms of address here and elsewhere. We should stick to "you", it's easier and more natural to read that way.
* I also shortened up some phrases I felt were overly qualified.

**Review by vvcephei — APPROVED** [review]
Thanks, @bbejeck.

As always, some wording nits are impossible to avoid, but it looks good to me overall!

**Comment by bbejeck** [comment]
\cc @mjsax

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

File: docs/streams/upgrade-guide.html (hunk 2 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

File: docs/streams/upgrade-guide.html (hunk 3 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

File: docs/streams/upgrade-guide.html (hunk 2 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

**Comment by bbejeck** [comment]
Thanks, @ableegoldman  and @vvcephei for the review! I've updated per comments.

File: docs/streams/upgrade-guide.html [hunk 4]
```diff
@@ -76,8 +76,13 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
...
+        join, the internal names of the state stores would shift, requiring an application reset when redeploying.  In the 2.4.0 release, Kafka Streams adds the <code>StreamJoined</code>
+        object, which gives users the ability to name the join processor, repartition topic(s) (if a repartition is required), and the state stores involved in the join.  Also, by naming the state stores, the changelog topics backing the state stores are named as well.
 
+        Another feature delivered by <code>StreamJoined</code> is that you can now configure the type of state store used in the join.  You can now elect to use in-memory stores or custom state stores for a stream-stream join.  Note that the provided stores will not be available for querying via Interactive Queries.  With the addition of <code>StreamJoined</code>, stream-stream join operations using <code>Joined</code> have been deprecated. Please switch over to stream-stream join methods using the new overloaded methods.  You can get more details from <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-479%3A+Add+StreamJoined+config+object+to+Join">KIP-479</a>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: Line too long...

File: docs/streams/upgrade-guide.html [hunk 5]
```diff
@@ -76,8 +76,13 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
 
     <!-- Placeholder KIP-213 -->
     <!-- Placeholder KIP-307 -->
-    <!-- Placeholder KIP-479 -->
+    <p>
+        Before the 2.4.0 version of Kafa Streams, users of the DSL could not name the state stores involved in a stream-stream join.  If users added DSL operations before of the 
+        join, the internal names of the state stores would shift, requiring an application reset when redeploying.  In the 2.4.0 release, Kafka Streams adds the <code>StreamJoined</code>
+        object, which gives users the ability to name the join processor, repartition topic(s) (if a repartition is required), and the state stores involved in the join.  Also, by naming the state stores, the changelog topics backing the state stores are named as well.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: Missing `<br /> ` markup ?

Comment (ableegoldman) [reply docs/streams/upgrade-guide.html]: Is it `<br>` or `<br />`? I've seen both in the docs and both seem to work, but what's the difference?

Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

File: docs/streams/upgrade-guide.html [hunk 6]
```diff
@@ -76,7 +76,14 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
 
     <!-- Placeholder KIP-213 -->
     <!-- Placeholder KIP-307 -->
-    <!-- Placeholder KIP-479 -->
+    <p>
+        Before the 2.4.0 version of Kafa Streams, users of the DSL could not name the state stores involved in a stream-stream join.  If users added DSL operations before of the 
+        join, the internal names of the state stores would shift, requiring an application reset when redeploying.  In the 2.4.0 release, Kafka Streams adds the <code>StreamJoined</code>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: `added DSL operations before of the join` -- sound confusing.

`If users changed their topology and added a operator before a join,`

File: docs/streams/upgrade-guide.html [hunk 7]
```diff
@@ -76,7 +76,14 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
...
+        object, which gives users the ability to name the join processor, repartition topic(s) (if a repartition is required), and the state stores involved in the join.  Also, by naming the state stores, the changelog topics backing the state stores are named as well.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: nit `object` -> `class`

Comment (mjsax) [reply docs/streams/upgrade-guide.html]: line is still quite long.

File: docs/streams/upgrade-guide.html [hunk 8]
```diff
@@ -76,7 +76,14 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
...
+        <br/>
+        Another feature delivered by <code>StreamJoined</code> is that you can now configure the type of state store used in the join.  You can now elect to use in-memory stores or custom state stores for a stream-stream join.  Note that the provided stores will not be available for querying via Interactive Queries.  With the addition of <code>StreamJoined</code>, stream-stream join operations using <code>Joined</code> have been deprecated. Please switch over to stream-stream join methods using the new overloaded methods.  You can get more details from
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: `You can now elect ` -> `You can elect` (the `now` is repetitive as it's used in the sentence before already)

Also, this line is too long

Comment (mjsax) [reply docs/streams/upgrade-guide.html]: `Note that the provided stores will not be available for querying via Interactive Queries` -> this is also true if the store in only named (might be worth to mention in the paragraph above already)?

File: docs/streams/upgrade-guide.html [hunk 9]
```diff
@@ -76,7 +76,14 @@ <h3><a id="streams_api_changes_240" href="#streams_api_changes_240">Streams API
...
+        join, the internal names of the state stores would shift, requiring an application reset when redeploying.  In the 2.4.0 release, Kafka Streams adds the <code>StreamJoined</code>
+        object, which gives users the ability to name the join processor, repartition topic(s) (if a repartition is required), and the state stores involved in the join.  Also, by naming the state stores, the changelog topics backing the state stores are named as well.
+        <br/>
+        Another feature delivered by <code>StreamJoined</code> is that you can now configure the type of state store used in the join.  You can now elect to use in-memory stores or custom state stores for a stream-stream join.  Note that the provided stores will not be available for querying via Interactive Queries.  With the addition of <code>StreamJoined</code>, stream-stream join operations using <code>Joined</code> have been deprecated. Please switch over to stream-stream join methods using the new overloaded methods.  You can get more details from 
+        <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-479%3A+Add+StreamJoined+config+object+to+Join">KIP-479</a>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: nit. Missing `.` at the end of the sentence

**Comment by mjsax** [comment]
@bbejeck This needs to be rebased to resolve merged conflicts, too.

File: docs/streams/upgrade-guide.html (hunk 6 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

File: docs/streams/upgrade-guide.html (hunk 7 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

File: docs/streams/upgrade-guide.html (hunk 9 above)
Comment (bbejeck) [reply docs/streams/upgrade-guide.html]: ack

**Comment by bbejeck** [comment]
@mjsax updated and rebased this

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.4`.

**Pull request merged**